WT-6463 History store operations should honor cache size

keitharnoldsmith opened the pull request:
The primary change here is that we no longer set `WT_SESSION_IGNORE_CACHE_SIZE` while a session has an open history store cursor, allowing the thread to perform eviction as normal.  There were two places where this caused trouble, necessitating additional changes.

1. An eviction thread (and possibly in other cases) that opens a history store cursor during eviction could attempt to re-enter eviction while using the cursor since `cursor_enter()` checks to see if eviction is needed.  The often leads to the eviction thread hanging.  To avoid this scenario, I have introduced a new session flag, `WT_SESSION_EVICTING` that I set when we start eviction (in `wt_evict()`).  When this flag is set, a session won't call into eviction. Mostly we can check for this in `wt_cache_eviction_check()`. The one exception (i.e., addition check) is when doing forced eviction in `wt_page_in_func()`.

 2. Eviction and reconciliation often needs to access the history store. This can now happen in a session that already has a history store cursor open. When we enter eviction (`wt_evict()`) we save the history store cursor (if any) so that if/when eviction needs a history store cursor, it will open a new one and not mess up (e.g., by resetting) the sessions existing hs cursor.

In making these changes, I noticed that the `WT_SESSION_HS_CURSOR` flag appears to be completely redundant. So I've removed it. For historical reasons, this flag was cleared during checkpoint, but in looking at the code (and testing) I couldn't find any benefit from this. 

This PR includes the changes from the currently unmerged PR for WT-6274.  

@luke-pearson: I noticed that you made a recent change to the forced eviction code in `wt_page_in_func()`.  That was one part of this PR where I wasn't entirely sure about undesirable side effects.  So I picked you as a reviewer.  Feel free to nominate someone else.

luke-pearson commented on src/evict/evict_lru.c:
```diff
@@ -284,10 +282,9 @@ __wt_evict_thread_run(WT_SESSION_IMPL *session, WT_THREAD *thread)
-        session_flags = 0; /* [-Werror=maybe-uninitialized] */
-        WT_RET(__wt_hs_cursor(session, &session_flags, &is_owner));
-        WT_RET(__wt_hs_cursor_close(session, session_flags, is_owner));
+    if (session->hs_cursor == NULL && !F_ISSET(conn, WT_CONN_IN_MEMORY | WT_CONN_READONLY)) {
```
Why the addition of WT_CONN_READONLY here?

keitharnoldsmith replied:
I had noticed elsewhere (e.g., `wt_hs_create()`) that we don't set up or use the history store if either `WT_CONN_IN_MEMORY` or `WT_CONN_READONLY` is set. So I updated the condition here to match.

sulabhM replied:
The reason is in-memory connection and read-only do not create a history store.

luke-pearson replied:
You could still be running in read only and attempt to read something from the history store in a previous iteration no?

keitharnoldsmith replied:
Ah.  Yes, of course.  I'll remove the WT_CONN_READONLY check here.

luke-pearson commented on src/btree/bt_read.c:
```diff
@@ -333,9 +333,11 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
+             * evicting something else.
              */
-            if (force_attempts < 10 && __evict_force_check(session, ref)) {
+            if (!F_ISSET(session, WT_SESSION_EVICTING) && force_attempts < 10 &&
```
This looks correct to me, we shouldn't attempt to evict while already evicting.

luke-pearson replied:
What would the call path be for us to arrive here while already evicting out of curiosity?

keitharnoldsmith replied:
> What would the call path be for us to arrive here while already evicting out of curiosity?

You're in luck. I still have a copy of that stack trace :-) 

I saw this during an eviction triggered from an eviction thread. During reconciliation it needed to insert a record in the history store and landed here when paging in the appropriate leaf node from the history store btree. As I recall the page's memory footprint was about 1MB which was sufficient to trigger `evict_force_check()`.

sulabhM replied:
Is that because when already evicting, we can only do a page read on a history store page. I think that is true. Maybe needs a comment around that.

luke-pearson commented on src/evict/evict_page.c:
```diff
@@ -8,10 +8,53 @@
+{
+    WT_ASSERT(session, F_ISSET(session, WT_SESSION_EVICTING));
+
+    /* If caller was using a history store cursor it should have closed it by now. */
```
Nit: If the caller

keitharnoldsmith replied:
Fixed.

luke-pearson commented on src/support/thread_group.c:
```diff
@@ -175,14 +175,10 @@ __thread_group_resize(WT_SESSION_IMPL *session, WT_THREAD_GROUP *group, uint32_t
-        /*
-         * Threads get their own session and hs table cursor (if the hs table is open).
-         */
+        /* Threads get their own session */
```
Should have a full stop at end of comment.

keitharnoldsmith replied:
Fixed.

keitharnoldsmith commented:
In addition to addressing @luke-pearson's comments, I merged up to current development branch, since it's been a while since the initial PR.  The Evergreen failures look like WT-6469.

agorrod commented:
@sulabhM please review this change.

sulabhM commented on src/evict/evict_page.c:
```diff
@@ -8,10 +8,53 @@
 static int __evict_page_dirty_update(WT_SESSION_IMPL *, WT_REF *, uint32_t);
 static int __evict_review(WT_SESSION_IMPL *, WT_REF *, uint32_t, bool *);
 
+/*
```
If I remember correctly we had made reconciliation non-reentrant for similar reasons, maybe couple of years back. We had issues with reconciliation repeatedly calling itself when using lookaside file. Are we sort of duplicating ```WT_SESSION_NO_RECONCILE``` with ```WT_SESSION_EVICTING```. Keith B. did that work, I am not sure if he is available for a review. @keitharnoldsmith can you have a look at ```WT_SESSION_NO_RECONCILE``` and see if it serves a different purpose than what you want to solve here.

keitharnoldsmith replied:
Thanks @sulabhM.  You raise a good point.  It seems like there is enough synergy here that we may be able to combine them in some way. I know that the current use of `WT_SESSION_NO_RECONCILE` isn't sufficient to prevent us from re-entering eviction, but it is worth spending a bit of time looking.  I'll see what I can come up with tomorrow.

sulabhM commented:
@agorrod , this pull req might change significantly as per the comments on WT-6463

keitharnoldsmith commented:
I've updated this PR with further (final?) changes based on @sulabhM's suggestion to see if we can take advantage of the similarity between the `WT_SESSION_NO_RECONCILE` and `WT_SESSION_EVICTING` flags.

It turns out that we can just re-use the existing `WT_SESSION_NO_RECONCILE` flag to avoid re-entering eviction. So I have changed the code to do that, and I've eliminated the `WT_SESSION_EVICTING` flag.

With this change, we also no longer need a check in `wt_page_in_func()` when trying to forcibly evict pages that are too big. In the problem cases, `WT_SESSION_NO_RECONCILE` is always set, which causes `wt_evict` to return `EBUSY` rather than doing eviction.

sulabhM commented on src/evict/evict_page.c:
```diff
@@ -192,6 +200,9 @@ __wt_evict(WT_SESSION_IMPL *session, WT_REF *ref, uint8_t previous_state, uint32
         goto done;
 
+    /* We're going ahead with eviction. We shouldn't get here if reconciliation isn't allowed. */
+    WT_ASSERT(session, !F_ISSET(session, WT_SESSION_NO_RECONCILE));
```
I looked at ```__evict_review()``` and we return ```EBUSY``` for eviction on dirty pages if ```WT_SESSION_NO_RECONCILE``` is set. We allow evicting clean pages with ```WT_SESSION_NO_RECONCILE``` set, as clean pages are just discarded and no reconciliation is needed. We can, in theory, hit this assert for clean pages.

Since we return EBUSY for the cases we care about, I think we do not need this assert anymore. I suggested adding an assert like that but after looking at ```__evict_review``` I think it is not needed,

keitharnoldsmith replied:
That makes sense.  I'll remove the assert.

sulabhM commented on src/include/cache.i:
```diff
@@ -458,6 +458,10 @@ __wt_cache_eviction_check(WT_SESSION_IMPL *session, bool busy, bool readonly, bo
         *didworkp = false;
 
+    /* Eviction causes reconciliation. So don't evict if we can't reconcile */
+    if (F_ISSET(session, WT_SESSION_NO_RECONCILE))
```
I am not 100% sure if allowing eviction of clean pages is relevant here. I will have to further look at the code. Earlier we would have successfully evicted clean pages and failed on dirty pages if no-reconciliation flag is set.

keitharnoldsmith replied:
@sulabhM : I thought about this a bit more after we talked.  I'm currently thinking this isn't a problem. Or, at least, it won't prevent any eviction we were doing before this PR.  Here's my reasoning:

The most common (only?) scenario where we call `wt_cache_eviction_check()` while we're in reconciliation is via `cursor_enter()` on history store cursor operations.  Before this PR, we would *never* evict on history store cursor operations because we would set `WT_SESSION_IGNORE_CACHE_SIZE` when opening a hs cursor.  

So even if we are missing opportunities here to evict clean pages.  We wouldn't have evicted those pages before this change.  

That still leaves a question of whether we really are missing opportunities to evict clean pages here, and if so how we could change the code to allow them.  But that strikes me as an opportunity for further optimization, not as a problem that this change would introduce.  

If that makes sense to you, I can merge this PR and create a new ticket to investigate/optimize clean page eviction here.  Or if I've missed something (quite possible; it's late here :-) I can go ahead with the original plan to ping @agorrod  for an opinion about this.

sulabhM replied:
This makes sense to me. We can create a separate ticket for further optimization. I think we can go-ahead with the merge.
